File: modules/example/linecount.c
/*
 * QSOL Example Module: Line Counter (C version)
 * A minimal, fast, and transparent line counting utility.
 */

#include <stdio.h>
#include <stdlib.h>

/* Count lines in a file stream */
long count_lines(FILE *fp) {
    long count = 0;
    int ch;
    int last_was_newline = 1;
    
    while ((ch = fgetc(fp)) != EOF) {
        if (ch == '\n') {
            count++;
            last_was_newline = 1;
        } else {
            last_was_newline = 0;
        }
    }
    
    /* Count last line if file doesn't end with newline */
    if (!last_was_newline) {
        count++;
    }
    
    return count;
}

/* Process a single file */
int process_file(const char *filename, long *total) {
    FILE *fp = fopen(filename, "r");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open '%s'\n", filename);
        return 1;
    }
    
    long count = count_lines(fp);
    printf("%s: %ld lines\n", filename, count);
    *total += count;
    
    fclose(fp);
    return 0;
}

int main(int argc, char *argv[]) {
    long total = 0;
    int errors = 0;
    
    if (argc < 2) {
        /* Read from stdin */
        total = count_lines(stdin);
        printf("%ld lines\n", total);
        return 0;
    }
    
    /* Process each file */
    for (int i = 1; i < argc; i++) {
        if (process_file(argv[i], &total) != 0) {
            errors++;
        }
    }
    
    /* Show total if multiple files */
    if (argc > 2) {
        printf("Total: %ld lines\n", total);
    }
    
    return errors > 0 ? 1 : 0;
}

File: modules/example/linecount.py
#!/usr/bin/env python3
"""
QSOL Example Module: Line Counter
A minimal, fast, and transparent line counting utility.
"""

import sys


def count_lines(filename):
    """Count lines in a file. Returns (line_count, error_message)."""
    try:
        with open(filename, 'r', encoding='utf-8', errors='ignore') as f:
            return sum(1 for _ in f), None
    except FileNotFoundError:
        return 0, f"Error: File '{filename}' not found"
    except PermissionError:
        return 0, f"Error: Permission denied for '{filename}'"
    except Exception as e:
        return 0, f"Error reading '{filename}': {e}"


def count_stdin():
    """Count lines from stdin."""
    return sum(1 for _ in sys.stdin)


def main():
    """Main entry point."""
    if len(sys.argv) < 2:
        # Read from stdin
        count = count_stdin()
        print(f"{count} lines")
        return 0
    
    # Read from files
    total = 0
    errors = []
    
    for filename in sys.argv[1:]:
        count, error = count_lines(filename)
        if error:
            errors.append(error)
        else:
            print(f"{filename}: {count} lines")
            total += count
    
    # Show total if multiple files
    if len(sys.argv) > 2:
        print(f"Total: {total} lines")
    
    # Print errors at the end
    for error in errors:
        print(error, file=sys.stderr)
    
    return 1 if errors else 0


if __name__ == "__main__":
    sys.exit(main())
